File: Android/app/src/main/java/com/olivierpicard/crachit/GameScene.java
package com.olivierpicard.crachit;

import android.graphics.Color;

import com.olivierpicard.crachit.AnimatedItem.Asteroid;
import com.olivierpicard.crachit.Graphics.GInterval;
import com.olivierpicard.crachit.Graphics.GLabel;
import com.olivierpicard.crachit.Graphics.GNode;
import com.olivierpicard.crachit.Graphics.GPoint;
import com.olivierpicard.crachit.Graphics.GScene;
import com.olivierpicard.crachit.Graphics.GSize;
import com.olivierpicard.crachit.Graphics.GTools;
import com.olivierpicard.crachit.Graphics.GVector;
import com.olivierpicard.crachit.Graphics.IGDrawable;
import com.olivierpicard.crachit.Shuttle.Shuttle;
import com.olivierpicard.crachit.Shuttle.ShuttleEnemiesGenerator;
import com.olivierpicard.crachit.Shuttle.ShuttleEnemy;
import com.olivierpicard.crachit.Shuttle.ShuttlePlayer;

import java.util.ArrayList;
import java.util.List;

import static com.olivierpicard.crachit.DataBaseHandler.*;

/**
 * Scène(vue) dans lequel le jeu prend vie
 * Created by olivierpicard on 03/04/2018.
 */

public class GameScene extends GScene {
    public enum GameState {
        PLAY,
        GAME_OVER,
        WELCOME,
        RESUME
    }

    private final float PLAYER_MOVING_SPEED = 3f;
    private GLabel score_label;
    private GameState state;
    private StarsGenerator starsGenerator_topLayer;
    private StarsGenerator starsGenerator_bottomLayer;
    private ShuttleEnemiesGenerator shuttle_enemy_generator;
    private AsteroidsGenerator asteroids_generator;
    private ShuttlePlayer player;
    private WelcomeScreen welcomeScreen;
    private GameOverScreen gameOver_screen;
    private TutoImage tutoImage;
    public static int scoreToResumeFrom = 0;
    public static List<DataBaseHandler.ItemRestaurationTable> itemsToResumeFrom = null;
    public volatile static GameState flag_stateToSwitchTo = null;


    public int getScore() {
        return Integer.valueOf(this.score_label.getText());
    }

    public void setScore(int score) {
        this.score_label.setText(Integer.toString(score));
    }


    public void didInitialized() {
        this.player = new ShuttlePlayer();
        this.player.setPosition(GTools.fromSceneToScreenPos(this.getSize(), new GPoint(0.5f, 0.2f)));
        this.starsGenerator_topLayer = new StarsGenerator(this, 8,
                new GInterval(60, 80),
                new GSize(2,3),
                1f);
        this.starsGenerator_bottomLayer = new StarsGenerator(this, 10,
                new GInterval(40, 60),
                new GSize(2,2),
                0.4f);
        this.shuttle_enemy_generator = new ShuttleEnemiesGenerator(this, player);

        this.asteroids_generator = new AsteroidsGenerator(this);

        this.score_label = new GLabel("0");
        this.score_label.setFontSize(55);
        this.score_label.setColor(Color.WHITE);
        this.score_label.setAlpha(125);
        this.score_label.setHidden(true);
        this.score_label.setPosition(GTools.fromSceneToScreenPos(this.getSize(), new GPoint(0.5f, 0.8f)));
        addChild(this.score_label);

        this.gameOver_screen = new GameOverScreen(this);
        this.welcomeScreen = new WelcomeScreen(this);
        if(this.flag_stateToSwitchTo != GameState.RESUME)
            flag_stateToSwitchTo = GameState.WELCOME;

    }


    public void start() {
        reset();


        this.welcomeScreen.hide();
        this.gameOver_screen.hide();
        this.state = GameState.PLAY;
        this.score_label.setHidden(false);
        this.player.direction = GVector.zero();
        this.asteroids_generator.enable = true;
        this.shuttle_enemy_generator.enable = true;
        this.player.lifeBar.setValue(this.player.stats.defense);
        this.player.setPosition(GTools.fromSceneToScreenPos(this.getSize(), new GPoint(0.5f, 0.2f)));
        this.addChild(player);

        if(this.flag_stateToSwitchTo != GameState.RESUME) {
            setScore(0);
            this.tutoImage = new TutoImage(this);
            return;
        }


        setScore(scoreToResumeFrom);
        if(itemsToResumeFrom == null) return;
        for(DataBaseHandler.ItemRestaurationTable item : itemsToResumeFrom) {
            if(item.classType.equals(ShuttleEnemy.class.getSimpleName()))
                this.shuttle_enemy_generator.restaure(item);
            else if(item.classType.equals(Asteroid.class.getSimpleName()))
                this.asteroids_generator.restaure(item);
            else if(item.classType.equals(ShuttlePlayer.class.getSimpleName()))
                this.player.setPosition(new GPoint(item.xPos, item.yPos));
        }
    }


    public void gameOver() {
        this.shuttle_enemy_generator.enable = false;
        this.asteroids_generator.enable = false;
        this.score_label.setHidden(true);
        this.gameOver_screen.savedItems = saveItem();
        flag_stateToSwitchTo = GameState.GAME_OVER;
        if(this.tutoImage != null) {
            this.tutoImage.delete();
            this.tutoImage = null;
        }
    }


    @Override
    public void update(long currentTime) {
        checkFlagGameStateChangeRequest();
        if(this.tutoImage == null) {
            this.shuttle_enemy_generator.generate(currentTime);
            this.asteroids_generator.generate(currentTime);
        }
        this.starsGenerator_topLayer.generate();
        this.starsGenerator_bottomLayer.generate();

        for(GNode node : children) {
            if(!(node instanceof MovingItem)) continue;
            MovingItem movingItem = (MovingItem) node;
            movingItem.update(currentTime);
            if(movingItem.isOutOfScreen())
                removeChild(node);
            else if(node instanceof ICollisionable) {
                final ICollisionable collisionableItem = (ICollisionable) node;
                final List<ICollisionable> overlapsedItems = overlapsListItems(collisionableItem);
                for(ICollisionable ovItem : overlapsedItems) {
                    collisionableItem.inCollisionWith(ovItem);
                    ovItem.inCollisionWith(collisionableItem);
                }
            }
        }
    }


    private List<ICollisionable> overlapsListItems(ICollisionable item) {
        List<ICollisionable> list = new ArrayList<>();
        for(GNode node : this.children) {
            if(!(node instanceof ICollisionable)) continue;
            final ICollisionable child = (ICollisionable) node;
            if(item == child) continue;
            if(item.isOverlaps(child)) list.add(child);
        }
        return list;
    }



    @Override
    public void touchDown(GPoint pos) {
        if(tutoImage != null) {
            this.tutoImage.delete();
            this.tutoImage = null;
        }
        if(this.state == GameState.PLAY) {
            if(pos.x < this.getSize().width/2){
                player.direction.dx = -PLAYER_MOVING_SPEED;
            } else if(pos.x > this.getSize().width/2) {
                player.direction.dx = PLAYER_MOVING_SPEED;
            }
        }
    }

    @Override
    public void touchUp(GPoint pos) {
        if(this.state == GameState.PLAY) player.direction.dx = 0;
        else if(this.state == GameState.WELCOME) this.welcomeScreen.touchUp(pos);
        else gameOver_screen.touchUp(pos);
    }


    private void checkFlagGameStateChangeRequest() {
        if(this.flag_stateToSwitchTo == null) return;

        switch (this.flag_stateToSwitchTo) {
            case PLAY:
            case RESUME:
                start();
                break;
            case WELCOME:
                gameOver_screen.hide();
                welcomeScreen.show();
                break;
            case GAME_OVER:
                welcomeScreen.hide();
                gameOver_screen.show();
                break;
        }
        if(this.flag_stateToSwitchTo == GameState.RESUME)
            this.flag_stateToSwitchTo = GameState.PLAY;
        this.state = flag_stateToSwitchTo;
        flag_stateToSwitchTo = null;
    }


    private void reset() {
        List<GNode> itemToDelete = new ArrayList<>();
        for(GNode node : this.children)
            if(node instanceof ICollisionable)
                itemToDelete.add(node);
        removeChildren(itemToDelete);
    }


    public List<DataBaseHandler.ItemRestaurationTable> saveItem() {
        List<DataBaseHandler.ItemRestaurationTable> items = new ArrayList<>();
        for(GNode node : this.children) {
            if(!(node instanceof Asteroid) && !(node instanceof Shuttle)) continue;
            MovingItem mv = (MovingItem)node;
            DataBaseHandler.ItemRestaurationTable item = DataBaseHandler.reference.new ItemRestaurationTable();
            item.xPos = mv.getPosition().x;
            item.yPos = mv.getPosition().y;
            item.dx = mv.direction.dx;
            item.dy = mv.direction.dy;
            item.zPosition = mv.getZPosition();
            item.zRotation = mv.getZRotation();
            if(node instanceof ShuttleEnemy) {
                ShuttleEnemy se = (ShuttleEnemy)node;
                item.option1 = se.bitmapName;
                item.classType = ShuttleEnemy.class.getSimpleName();
                item.life = se.lifeBar.getValue();
            } else if(node instanceof Asteroid)
                item.classType = Asteroid.class.getSimpleName();
            else if(node instanceof ShuttlePlayer)
                item.classType = ShuttlePlayer.class.getSimpleName();

            items.add(item);
        }
        return items;
    }





}

File: iOS/Crash it/GameScene.swift
//
//  GameScene.swift
//  Crash it
//
//  Created by Olivier Picard on 11/03/2018.
//  Copyright © 2018 Olivier Picard. All rights reserved.
//

import SpriteKit
import GameplayKit

class GameScene: SKScene {
    enum GameState {
        case play
        case gameOver
        case welcome
    }
    
    class TutoImage {
        var click_left: SKSpriteNode
        var click_right: SKSpriteNode
        var middle_line: SKSpriteNode
        var explain_text: SKLabelNode
        let scene: SKScene

        
        init(_ scene: SKScene) {
           
            self.scene = scene
            self.middle_line = SKSpriteNode(color: UIColor.white, size: CGSize(width: 2, height: scene.size.height))
            self.middle_line.position = Tools.fromSceneToWorldPosition(screenSpacePos: CGPoint(x: 0.5, y: 0.5))
            
            self.click_left = SKSpriteNode(texture: SKTexture(image: #imageLiteral(resourceName: "pointer")), color: UIColor.white, size: CGSize(width: 60, height: 60))
            self.click_left.position = Tools.fromSceneToWorldPosition(screenSpacePos: CGPoint(x: 0.25, y: 0.3))
            
            self.click_right = SKSpriteNode(texture: SKTexture(image: #imageLiteral(resourceName: "pointer")), color: UIColor.white, size: CGSize(width: 70, height: 70))
            self.click_right.position = Tools.fromSceneToWorldPosition(screenSpacePos: CGPoint(x: 0.75, y: 0.4))
            
            self.explain_text = SKLabelNode(text: "Cliquez à droite ou à gauche de l'écran\nPour déplacer le vaisseau")
            self.explain_text.numberOfLines = 0
            self.explain_text.horizontalAlignmentMode = .center
            self.explain_text.preferredMaxLayoutWidth = scene.size.width - 40
            self.explain_text.alpha = 0.7
            self.explain_text.fontName = "HelveticaNeue-Light"
            self.explain_text.fontSize = 15
            self.explain_text.fontColor = UIColor.white
            self.explain_text.position = Tools.fromSceneToWorldPosition(screenSpacePos: CGPoint(x: 0.5, y: 0.6))
            
            click_left.alpha = 0.5
            click_right.alpha = 0.5
            middle_line.alpha = 0.35
            
            self.scene.addChild(self.click_left)
            self.scene.addChild(self.click_right)
            self.scene.addChild(self.middle_line)
            self.scene.addChild(self.explain_text)
        }
        
        deinit {
            self.scene.removeChildren(in: [self.click_left, self.click_right, self.middle_line, self.explain_text])
        }
    }
    
    
    
    private let PLAYER_MOVING_SPEED:CGFloat = 3
    private var player: ShuttlePlayer!
    internal var score_label: SKLabelNode!
    private var starsGenerator_topLayer:StarsGenerator!
    private var starsGenerator_bottomLayer:StarsGenerator!
    private var shuttle_enemy_generator: ShuttleEnemyGenerator!
    private var asteroids_generator: AsteroidsGenerator!
    private var gameOver_screen: GameOverScreen!
    private var welcome_screen: WelcomeScreen!
    private var state: GameState!
    internal var view_Controller: GameViewController!
    private var tutoImage: TutoImage?
    
    
    
    
    var score: Int {
        get { return Int(self.score_label.text!)! }
        set {
            score_label.text! = String(newValue)
        }
    }
    
    
    override func didMove(to view: SKView) {
//        UserDefaults.standard.removeObject(forKey: Tools.KEY_DEFAULT_GAME_ITEMS)
//        UserDefaults.standard.removeObject(forKey: Tools.KEY_DEFAULT_GAMEINFOS)
//        UserDefaults.standard.removeObject(forKey: Tools.KEY_DEFAULT_SCORES)
        Tools.scene_size = self.size
        self.score_label = SKLabelNode.init(text: "0")
        self.score_label.alpha = 0.5
        self.score_label.fontName = "HelveticaNeue-Light"
        self.score_label.fontSize = 55
        self.score_label.fontColor = UIColor.white
        self.score_label.position = Tools.fromSceneToWorldPosition(screenSpacePos: CGPoint(x: 0.5, y: 0.8))
        self.score_label.isHidden = true
        self.addChild(self.score_label)
        
        self.player = ShuttlePlayer()
        self.player.position = Tools.fromSceneToWorldPosition(screenSpacePos: CGPoint(x: 0.5, y: 0.2))
        
        self.starsGenerator_topLayer = StarsGenerator(scene: self,
                                                      screenSize: self.size,
                                                      starsPercent: 8,
                                                      opacityRange: Tools.Interval(min: 50, max: 80),
                                                      starSize: CGSize(width: 1, height: 2),
                                                      speedFactor: 1.0)
        
        self.starsGenerator_bottomLayer = StarsGenerator(scene: self,
                                                         screenSize: self.size,
                                                         starsPercent: 13,
                                                         opacityRange: Tools.Interval(min: 40, max: 60),
                                                         starSize: CGSize(width: 1, height: 1),
                                                         speedFactor: 0.4)
        
        self.shuttle_enemy_generator = ShuttleEnemyGenerator(scene: self, target: player)
        self.asteroids_generator = AsteroidsGenerator(scene: self)

        self.gameOver_screen = GameOverScreen(scene: self)
        self.welcome_screen = WelcomeScreen(scene: self)
        
        if(self.view_Controller.initWithScore != nil) { start(); return }
        
        self.welcome_screen.show()
        self.state = GameState.welcome
        
    }
    
    func start() {
        reset()
        self.score = self.view_Controller.initWithScore ?? 0
        self.view_Controller.initWithScore = nil
        self.tutoImage = TutoImage(self)
        self.addChild(player)
        self.state = GameState.play
        self.score_label.isHidden = false
        self.player.direction = CGVector.zero
        self.asteroids_generator.enable = true
        self.shuttle_enemy_generator.enable = true
        self.player.lifeBar.value = self.player.stats.defense
        self.player.position = Tools.fromSceneToWorldPosition(
            screenSpacePos: CGPoint(x: 0.5, y: 0.2))
        
        guard let itemsToRestaure = self.view_Controller.initWithItems else {return}
        
        for itemToRestaure in itemsToRestaure {
            if itemToRestaure.type == "Asteroid" {
                self.asteroids_generator.restaure(item: itemToRestaure)
            }
            else if itemToRestaure.type == "ShuttleEnemy" {
                self.shuttle_enemy_generator.restaure(item: itemToRestaure)
            }
            else if itemToRestaure.type == "ShuttlePlayer" {
                self.player.position = itemToRestaure.position!
            }
        }
    }
    
    
    
    func touchDown(atPoint pos : CGPoint) {
        self.tutoImage = nil
        if(self.state == GameState.play) {
            if(pos.x < self.size.width/2){
                player.direction.dx = -self.PLAYER_MOVING_SPEED
            } else if(pos.x > self.size.width/2) {
                player.direction.dx = self.PLAYER_MOVING_SPEED
            }
        }
    }
    
    func touchMoved(toPoint pos : CGPoint) {
        if self.state == GameState.play {
            
        }
    }
    
    func touchUp(atPoint pos : CGPoint) {
        if self.state == GameState.play { player.direction.dx = 0 }
        else if self.state == GameState.welcome { self.welcome_screen.touchUp(pos) }
        else { gameOver_screen.touchUp(pos) }
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        for t in touches { self.touchDown(atPoint: t.location(in: self)) }
    }
    
    
    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        for t in touches { self.touchUp(atPoint: t.location(in: self)) }
    }
    
    override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
        for t in touches { self.touchUp(atPoint: t.location(in: self)) }
    }
    
    
    override func update(_ currentTime: TimeInterval) {
        self.starsGenerator_topLayer.generate()
        self.starsGenerator_bottomLayer.generate()
        
        if self.tutoImage == nil {
            self.shuttle_enemy_generator.generate(currentTime)
            self.asteroids_generator.generate(currentTime)
        }
        
        var itemsToDelete:[SKNode] = []
        for item in self.children {
            if let movingItem = item as? MovingItem {
                movingItem.update(currentTime)
                if(movingItem.isOutOfScreen(screenSize: self.size)) {
                    itemsToDelete.append(item)
                }
                else if let collisionableItem = item as? Collisionable {
                    let overlapsedItems = overlapsListItems(collisionableItem)
                    for ovItem in overlapsedItems {
                        collisionableItem.inCollisionWith(item: ovItem)
                        ovItem.inCollisionWith(item: collisionableItem)
                    }
                }
            }
        }
        self.removeChildren(in: itemsToDelete)
    }
    
    
    func overlapsListItems(_ item: Collisionable) -> [Collisionable] {
        var list: [Collisionable] = []
        for i in self.children {
            if let child = i as? Collisionable {
                if(item as! MovingItem == child as! MovingItem) {continue}
                if item.isOverlaps(child) { list.append(child) }
            }
        }
        return list
    }
    
    
    
    func gameOver() {
        self.shuttle_enemy_generator.enable = false
        self.asteroids_generator.enable = false
        self.score_label.isHidden = true
        self.state = GameState.gameOver
        self.gameOver_screen.show()
        self.tutoImage = nil
    }
    
    
    func saveItems() -> [Tools.ItemConf] {
        var items: [Tools.ItemConf] = []
        for child in self.children {
            if !(child is Shuttle) && !(child is Asteroid) {continue}
            
            let position = CGPoint.init(x: child.position.x, y: child.position.y)
            let zPosition = CGFloat(child.zPosition)
            let zRotation = CGFloat(child.zRotation)
            let movingItem = (child as! MovingItem)
            let direction = CGVector(dx: movingItem.direction.dx, dy: movingItem.direction.dy)
            var image: String?
            var type: String?
            var life: Int?
            if let se = child as? ShuttleEnemy {
                image = se.image.accessibilityIdentifier
                type = "ShuttleEnemy"
                life = Int(se.lifeBar.value)
            } else if let ast = child as? Asteroid {
                type = "Asteroid"
            } else if let play = child as? ShuttlePlayer {
                type = "ShuttlePlayer"
            }
            items.append(Tools.ItemConf.init(type, position, image, zPosition, zRotation, life, direction))
        }
        return items
    }
    
    
    
    func switchScreen(_ state: GameState) {
        if state == GameState.gameOver {
            self.state = state
            self.gameOver_screen.show()
        } else if state == GameState.welcome {
            self.state = state
            self.welcome_screen.show()
        }
    }
    
    func reset() {
        var nodesToDelete: [SKNode] = []
        for child in self.children {
            if !(child is Collisionable) {continue}
            nodesToDelete.append(child)
        }
        self.removeChildren(in: nodesToDelete)
    }
}





